File: client/src/api/layouts.ts
import { request } from "./client";
import type { components } from "./types";

type LayoutCreate = components["schemas"]["LayoutCreate"];
type LayoutRead = components["schemas"]["LayoutRead"];
type LayoutUpdate = components["schemas"]["LayoutUpdate"];

export function listLayouts() {
	return request<LayoutRead[]>("/layouts");
}

export function getLayout(id: string) {
	return request<LayoutRead>(`/layouts/${id}`);
}

export function createLayout(payload: LayoutCreate) {
	return request<LayoutRead>("/layouts", {
		method: "POST",
		body: JSON.stringify(payload),
	});
}

export function updateLayout(id: string, payload: LayoutUpdate) {
	return request<LayoutRead>(`/layouts/${id}`, {
		method: "PATCH",
		body: JSON.stringify(payload),
	});
}

export function deleteLayout(id: string) {
	return request<void>(`/layouts/${id}`, {
		method: "DELETE",
	});
}

File: api/core/routes/layouts.py
from uuid import UUID
from django.shortcuts import get_object_or_404
from django.utils import timezone
from ninja import Router
from ninja.responses import Response

from core.models import Layout
from core.schemas.layout import LayoutCreate, LayoutRead, LayoutUpdate

router = Router(tags=["layouts"])


@router.post("/layouts", response=LayoutRead)
def create_layout(request, payload: LayoutCreate):
    layout = Layout.objects.create(
        name=payload.name,
        content=payload.content,
    )
    return layout


@router.get("/layouts", response=list[LayoutRead])
def list_layouts(request):
    return list(Layout.objects.all())


@router.get("/layouts/{layout_id}", response=LayoutRead)
def get_layout(request, layout_id: UUID):
    layout = get_object_or_404(Layout, id=layout_id)
    return layout


@router.patch("/layouts/{layout_id}", response=LayoutRead)
def update_layout(request, layout_id: UUID, payload: LayoutUpdate):
    layout = get_object_or_404(Layout, id=layout_id)

    if payload.name is not None:
        layout.name = payload.name
    if payload.content is not None:
        layout.content = payload.content

    layout.save()
    return layout


@router.delete("/layouts/{layout_id}", response={204: None})
def delete_layout(request, layout_id: UUID):
    layout = get_object_or_404(Layout, id=layout_id)
    layout.delete()
    return Response(status=204)
